commons-core/src/main/scala/com/avsystem/commons/serialization/json/JsonStringInput.scala

```diff
@@ -43,24 +43,84 @@ class JsonStringInput(
     if (reader.jsonType != jsonType) expectedError(jsonType)
     else reader.currentValue.asInstanceOf[T]
 
-  private def matchNumericString[T](toNumber: String => T): T =
-    if (reader.jsonType == JsonType.number || reader.jsonType == JsonType.string) {
-      val str = reader.currentValue.asInstanceOf[String]
-      try toNumber(str) catch {
-        case e: NumberFormatException =>
-          throw new ParseException(s"Invalid number format: $str ${reader.posInfo(startIdx)}", e)
-      }
-    } else expectedError(JsonType.number)
-
   def readNull(): Boolean = reader.jsonType == JsonType.`null`
   def readString(): String = checkedValue[String](JsonType.string)
   def readBoolean(): Boolean = checkedValue[Boolean](JsonType.boolean)
-  def readInt(): Int = matchNumericString(_.toInt)
-  def readLong(): Long = matchNumericString(_.toLong)
-  override def readFloat(): Float = matchNumericString(_.toFloat)
-  def readDouble(): Double = matchNumericString(_.toDouble)
-  def readBigInt(): BigInt = matchNumericString(BigInt(_))
-  def readBigDecimal(): BigDecimal = matchNumericString(BigDecimal(_, options.mathContext))
+
+  private def isInteger(str: String): Boolean = {
+    @tailrec def loop(i: Int): Boolean =
+      i >= str.length || str.charAt(i).isDigit && loop(i + 1)
+    str.nonEmpty && {
+      if (str.charAt(0) == '-') str.length > 1 && loop(1)
+      else loop(0)
+    }
+  }
+
+  private def parseNumber[T](parse: String => T): T = {
+    val str = reader.jsonType match {
+      // accepting JSON strings as numbers (also because they may be Infinity/-Infinity/NaN)
+      case JsonType.number | JsonType.string => reader.currentValue.asInstanceOf[String]
+      case _ => expectedError(JsonType.number)
+    }
+    try parse(str) catch {
+      case e: NumberFormatException =>
+        throw new ParseException(s"Invalid number format: $str ${reader.posInfo(startIdx)}", e)
+    }
+  }
+
+  override def readByte(): Byte = parseNumber { str =>
+    if (isInteger(str)) str.toByte
+    else {
+      val dbl = str.toDouble
+      if (dbl.isValidByte) dbl.toByte
+      else throw new NumberFormatException(str)
+    }
+  }
+
+  override def readShort(): Short = parseNumber { str =>
+    if (isInteger(str)) str.toShort
+    else {
+      val dbl = str.toDouble
+      if (dbl.isValidShort) dbl.toShort
+      else throw new NumberFormatException(str)
+    }
+  }
+
+  def readInt(): Int = parseNumber { str =>
+    if (isInteger(str)) str.toInt
+    else {
+      val dbl = str.toDouble
+      if (dbl.isValidInt) dbl.toInt
+      else throw new NumberFormatException(str)
+    }
+  }
+
+  def readLong(): Long = parseNumber { str =>
+    if (isInteger(str)) str.toLong
+    else {
+      val bd = BigDecimal(str, options.mathContext)
```

Comment:
Using `BigDecimal` for parsing of `Long` values opens ability for DoS/DoW attacks which exploit O(n^2) complexity of Java's way to parse big numbers from textual representations.

[I've raised an issue for `BigInt` and `BigDecimal` types](https://github.com/AVSystem/scala-commons/issues/88) but now it goes to more widely used types.

Reply:
@plokhotnyuk do you have a safe `BigInt`/`BigDecimal` parsing implementation that we could lend?

Reply:
[Here](https://github.com/plokhotnyuk/jsoniter-scala/blob/6e3fa06217e18411b85dead678b48c4a2eef8a8e/jsoniter-scala-core/jvm/src/main/scala/com/github/plokhotnyuk/jsoniter_scala/core/JsonReader.scala#L1585-L1841) is the code which has O(n^1.5) complexity and checks limits for the scale and number of digits. I've used [the following default values](https://github.com/plokhotnyuk/jsoniter-scala/blob/b2fef1806987828eb256ec712e354a546d5e2a00/jsoniter-scala-macros/shared/src/main/scala/com/github/plokhotnyuk/jsoniter_scala/macros/JsonCodecMaker.scala#L59-L63) for them.